### PR TITLE
Category

### PR DESCRIPTION
**Pull request by INR4GE**

### CODE REVIEW

File: modules/user/user.service.js [hunk 1]
```diff
@@ -136,7 +136,7 @@ class UserService {
       throw new UserInputError(WRONG_CREDENTIALS, { statusCode: 400 });
     }
 
-    const token = generateToken(user._id, user.email);
+    const token = generateToken(user._id, user.email, { EXPIRES_IN: '10h' });
```
Comment (Alan18081) [inline modules/user/user.service.js:139]: Why do you need this change as a part of merge request?

File: typeDefs.js [hunk 2]
```diff
@@ -190,6 +190,7 @@ const typeDefs = gql`
 
     getAllCategories: [Category]
     getCategoryById(id: ID): CategoryResult
+    getSubcategories(id: ID!): [Category]
```
Comment (Alan18081) [inline typeDefs.js:193]: Please, rename param `id` to the `parentCategoryId` to improve readability and understandability of code

File: modules/category/category.graphql.js [hunk 3]
```diff
@@ -12,7 +12,7 @@ const categoryType = `
 
 const categoryInput = `
     input CategoryInput {
-    categoryCode: String
+    code: String
     name: [LanguageInput]
     images: ImageSetInput
     subcategories: [ID]
```
Comment (Alan18081) [inline modules/category/category.graphql.js:18]: I thought that when we create a category we are going to provide subcategories as an array of objects with category data

File: modules/category/category.service.js [hunk 4]
```diff
@@ -22,39 +24,87 @@ class CategoryService {
...
     });
     return categoriesCount > 0;
   }
+
+  async getSubcategories(id) {
+    const category = await this.getCategoryById(id);
+    if (!category.isMain) {
+      return [];
+    }
+    const idList = category.subcategories.length
```
Comment (Alan18081) [inline modules/category/category.service.js:103]: If category doesn't have any subcategories it is much easier to just return empty array without any access to the database

File: modules/category/category.service.js [hunk 5]
```diff
@@ -22,39 +24,87 @@ class CategoryService {
...
+      ? category.subcategories
+      : [''];
```
Comment (Alan18081) [inline modules/category/category.service.js:105]: Generally, It is an invalid value for the `$in` operator, so don't use it because it can cause really strange issues

File: modules/category/category.graphql.js [hunk 6]
```diff
@@ -1,7 +1,7 @@
 const categoryType = `
     type Category {
     _id: ID!
-    categoryCode: String
+    code: String
```
[the remark this answers is not in the document]
Comment (INR4GE) [reply modules/category/category.graphql.js:4]: It was "code" in the Category model, had to fix it in .graphql to correspond each other

File: modules/category/category.service.js [hunk 7]
```diff
@@ -22,39 +24,87 @@ class CategoryService {
     if (!categoryToUpdate) {
       throw new Error(CATEGORY_NOT_FOUND);
     }
-    if (await this.checkCategoryExist(category, id)) {
```
[the remark this answers is not in the document]
Comment (INR4GE) [reply modules/category/category.service.js:25]: Fixed

File: modules/category/category.resolver.js [hunk 8]
```diff
@@ -16,12 +16,23 @@ const categoryQuery = {
       };
     }
   },
+  getSubcategories: async (parent, args) => {
+    try {
+      return await categoryService.getSubcategories(args.id);
+    } catch (e) {
+      console.log(e.message);
```
[the remark this answers is not in the document]
Comment (INR4GE) [reply modules/category/category.resolver.js:23]: removed

File: modules/category/category.service.js [hunk 9]
```diff
@@ -22,39 +24,87 @@ class CategoryService {
...
-      throw new Error(CATEGORY_ALREADY_EXIST);
-    }
     return await Category.findByIdAndUpdate(id, category, {
       new: true,
     });
   }
 
-  async addCategory(data) {
+  async addCategory(data, parentId) {
+    if (data.name.length < 2 || !data.code || (!data.isMain && !parentId)) {
```
[the remark this answers is not in the document]
Comment (INR4GE) [reply modules/category/category.service.js:33]: We do not always need parentID, because main categories dont have them, so they are nullable.
Separated errors.

File: modules/category/category.service.js [hunk 10]
```diff
@@ -22,39 +24,87 @@ class CategoryService {
...
+    if (!parentId) {
+      selectedId = null;
+    }
+    const parentCategory = await Category.findOne({ _id: selectedId });
+    const newCategory = new Category(data);
+
+    if (selectedId && !parentCategory) {
+      throw new Error(CATEGORY_NOT_FOUND);
+    }
+    const savedCategory = await newCategory.save();
```
[the remark this answers is not in the document]
Comment (INR4GE) [reply modules/category/category.service.js:49]: I need to do so, in order to save newly created category _id to parent.subcategories array. If i dont save it first, i won't get _id

File: modules/category/category.service.js [hunk 11]
```diff
@@ -22,39 +24,87 @@ class CategoryService {
...
+    }
     if (await this.checkCategoryExist(data)) {
       throw new Error(CATEGORY_ALREADY_EXIST);
     }
-    return new Category(data).save();
+    let selectedId = parentId;
+    if (!parentId) {
+      selectedId = null;
+    }
+    const parentCategory = await Category.findOne({ _id: selectedId });
```
[the remark this answers is not in the document]
Comment (INR4GE) [reply modules/category/category.service.js:43]: replaced

File: modules/user/user.service.js (hunk 1 above)
Comment (INR4GE) [reply modules/user/user.service.js:139]: There were some issues with token expiring too fast, so i just set it for admin to be longer. Removed.

File: typeDefs.js (hunk 2 above)
Comment (INR4GE) [reply typeDefs.js:193]: renamed

File: modules/category/category.graphql.js (hunk 3 above)
Comment (INR4GE) [reply modules/category/category.graphql.js:18]: The UI would become more complicated if we were to create subcategories simultaneously with main category, like more tabs embedded into other tabs. I think creating category first and then subcategories after would be much simpler.

File: modules/category/category.service.js (hunk 5 above)
Comment (INR4GE) [reply modules/category/category.service.js:105]: not providing an array will be crashing the operation, because it is required for $in operation to have an array

File: modules/category/category.service.js (hunk 4 above)
Comment (INR4GE) [reply modules/category/category.service.js:103]: Added check for subcategories

File: error-messages/category.messages.js [hunk 12]
```diff
@@ -2,10 +2,14 @@ const CATEGORY_NOT_FOUND = 'CATEGORY_NOT_FOUND';
 const CATEGORY_ALREADY_EXIST = 'CATEGORY_ALREADY_EXIST';
 const CATEGORY_IS_NOT_MAIN = 'CATEGORY_IS_NOT_MAIN';
 const WRONG_CATEGORY_DATA = 'WRONG_CATEGORY_DATA';
+const CATEGORY_INPUT_ERROR = 'CATEGORY_INPUT_ERROR';
+const PARENT_ID_NOT_PROVIDED = 'PARENT_ID_NOT_PROVIDED';
```
Comment (Alan18081) [inline error-messages/category.messages.js:6]: Please, rename the error message to `PARENT_CATEGORY_ID_NOT_PROVIDED`, I think it would be much more clear

File: error-messages/category.messages.js [hunk 13]
```diff
@@ -2,10 +2,14 @@ const CATEGORY_NOT_FOUND = 'CATEGORY_NOT_FOUND';
 const CATEGORY_ALREADY_EXIST = 'CATEGORY_ALREADY_EXIST';
 const CATEGORY_IS_NOT_MAIN = 'CATEGORY_IS_NOT_MAIN';
 const WRONG_CATEGORY_DATA = 'WRONG_CATEGORY_DATA';
+const CATEGORY_INPUT_ERROR = 'CATEGORY_INPUT_ERROR';
```
Comment (Alan18081) [inline error-messages/category.messages.js:5]: Please, make this message more meaningful because currently, it says nothing about the root cause of error

File: modules/category/category.service.js [hunk 14]
```diff
@@ -30,22 +35,18 @@ class CategoryService {
   }
 
   async addCategory(data, parentId) {
-    if (data.name.length < 2 || !data.code || (!data.isMain && !parentId)) {
-      throw new Error(WRONG_CATEGORY_DATA);
+    if (data.name.length < 2) {
+      throw new Error(CATEGORY_INPUT_ERROR);
```
Comment (Alan18081) [inline modules/category/category.service.js:39]: Please, make this error more meaningful and use `joi-validator` here

File: modules/category/category.service.js [hunk 15]
```diff
@@ -30,22 +35,18 @@ class CategoryService {
...
+    }
+    if (!data.isMain && !parentId) {
+      throw new Error(PARENT_ID_NOT_PROVIDED);
```
Comment (Alan18081) [inline modules/category/category.service.js:42]: Please, make this error more meaningful and use `joi-validator` here

File: modules/category/category.service.js [hunk 16]
```diff
@@ -21,6 +23,9 @@ class CategoryService {
 
   async updateCategory(id, category) {
     const categoryToUpdate = await Category.findById(id);
+    if (await this.checkCategoryExist(category, id)) {
```
Comment (Alan18081) [inline modules/category/category.service.js:26]: This check should be after the check for existence of category of id

File: modules/product/product.service.js [hunk 17]
```diff
@@ -15,13 +15,13 @@ class ProductsService {
   }
 
   getModelsByCategory(id) {
-    return Model.find({ category: id });
+    return Product.find({ category: id });
```
Comment (Alan18081) [inline modules/product/product.service.js:18]: Wait, should it be `Product` or `Model`

File: modules/product/product.service.js [hunk 18]
```diff
@@ -56,11 +56,10 @@ class ProductsService {
       };
     }
     if (price && price.length) {
-      const currencySign =
-      currency === 0 ? 'UAH' : currency === 1 ? 'USD' : '';
+      const currencySign =      currency === 0 ? 'UAH' : currency === 1 ? 'USD' : '';
```
Comment (Alan18081) [inline modules/product/product.service.js:59]: Please, validate currency values via `joi-validator`

File: modules/user/user.service.js [hunk 19]
```diff
@@ -136,7 +136,7 @@ class UserService {
       throw new UserInputError(WRONG_CREDENTIALS, { statusCode: 400 });
     }
 
-    const token = generateToken(user._id, user.email, { EXPIRES_IN: '10h' });
+    const token = generateToken(user._id, user.email);
```
Comment (Alan18081) [inline modules/user/user.service.js]: Please, revert this change, the token should expire

File: modules/product/product.service.js (hunk 17 above)
Comment (INR4GE) [reply modules/product/product.service.js:18]: I couldn't commit cause of this error, someone imported Product from './product.model', but used Model, that was not defined

File: modules/user/user.service.js (hunk 19 above)
Comment (INR4GE) [reply modules/user/user.service.js]: reverted

File: modules/product/product.service.js (hunk 18 above)
Comment (INR4GE) [reply modules/product/product.service.js:59]: sry, not mine component

File: modules/user/user.service.js [hunk 20]
```diff
@@ -222,9 +222,7 @@ class UserService {
       html: confirmationMessage(firstName, token, language),
     };
 
-    if (process.env.NODE_ENV !== 'test') {
-      await sendEmail(message);
-    }
+    if (process.env.NODE_ENV !== 'test') await sendEmail(message);
```
[the remark this answers is not in the document]
Comment (INR4GE) [reply modules/user/user.service.js:225]: reverted

File: error-messages/category.messages.js (hunk 12 above)
Comment (INR4GE) [reply error-messages/category.messages.js:6]: renamed

File: modules/category/category.service.js (hunk 14 above)
Comment (INR4GE) [reply modules/category/category.service.js:39]: used
